### ISSUE TITLE
Support "Exchanger" synchronization point

### ISSUE THREAD

**Issue by khovanskiy** [Kotlin/kotlinx.coroutines#1911]
Docs: https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/Exchanger.html

It will be useful for swapping values between coroutines.

Probably it can not be done with the `Channel` class. If I am wrong please correct me.

I suggest the following draft solution: 
```
class Exchanger<V> {
    private val holder = AtomicReference<Pair<V, CancellableContinuation<V>>>(null)

    suspend fun exchange(x: V): V = suspendCancellableCoroutine sc@{ cont ->
        var cache: Pair<V, CancellableContinuation<V>>? = null
        try_again@ while (true) {
            val current = holder.get()
            if (current == null) {
                if (cache == null) {
                    cache = Pair(x, cont)
                }
                if (!holder.compareAndSet(null, cache)) {
                    continue@try_again
                }
                break
            } else {
                if (!holder.compareAndSet(current, null)) {
                    continue@try_again
                }
                current.second.resume(x)
                cont.resume(current.first)
                break
            }
        }
    }
}
```

**Comment by qwwdfsad** [comment]
Could you please elaborate on the use of such primitive? E.g. what kind of widespread domain-specific problem it is supposed to solve?

**Comment by khovanskiy** [comment]
It provides a simple way to make two unidirectional connections to be bidirectional. Also, it useful then we have an asynchronous API with callbacks and want to use it like synchronous API.

For example:

1. Several JMS queues when the `correlationId` should be dynamically calculated based on message;
2. Some API for public social networks use the REST for action initialization and the result of action returns to callback URL, which usually is pooled in another thread and we have to generate the response for this callback endpoint;

**Comment by elizarov** [comment]
@khovanskiy Sorry, but I do not follow how the exchanger helps in those use-cases. Can you elaborate your use-cases a bit with code examples, please?

**Comment by khovanskiy** [comment]
The example.
Setup in configuration a single incoming message listener (like HTTP server or JMS queue):
```
val queue = Queue()
val correlationMap = mutableMapOf<String, Exchanger<Message>>()
queue.addListener { (correlationId, incomingRequest) ->
    // actually the incomingRequest is the response for outgoingRequest
    val exchanger = correlationMap.computeIfAbsent(correlationId) { Exchanger<Message>() }
    val outgoingResponse = exchange.exchange(incomingRequest)
    return@addListener outgoingResponse
}
```

A unidirectional send (the response will be asynchronously received only via listener):
```
val outgoingRequest = Message(...)
val correlationId: String = queue.send(outgoingRequest)
```
A bidirectional exchanging between the outgoing response and the incoming response:
```
val exchanger = correlationMap.computeIfAbsent(correlationId) { Exchanger() }
val outgoingResponse = Message(...)
val reply = exchanger.exchange(outgoingResponse) 
return reply
```

**Comment by fvasco** [comment]
Hi, @khovanskiy,
where did you seen applied this solution?

You wrote about two responses, these implies two simultaneous requests?
In your use case, you have to await a outcoming response before receive the incoming one?

**Comment by elizarov** [comment]
@khovanskiy Why wouldn't you use a coroutine and a pair of channels for each `correlationId`? I don't grasp the significance of an exchanger for this particular use-case. Why it has to be synchronous?

**Comment by khovanskiy** [comment]
Thank everyone for the help! 

I think the issue can be closed if you want. Now I can't provide a use-case because I solved my problem in another way.

In my case, each `correlationId` is the unique value that can be the same in only one listener and one sender. So performance tests showed that it is not the best choice to create a pair of channels for each `correlationId`. Eventually, I implemented a rather complex data structure based on `ConcurrentLinkedQueue` of `CompletableDeferred`. It ensured a strict order of messages and avoided extra suspends in most of the cases despite the linear search.
